File: src/lib/portfolio-analysis.js
// 组合分析的纯计算逻辑。原先内联在 Portfolio.jsx（约 250 行），抽出后可单测。
import { formatPercent, formatSignedPct } from "./format.js";

export const CHART_COLORS = ["#2563eb", "#0f766e", "#dc2626", "#7c3aed", "#d97706", "#0891b2", "#64748b", "#be185d"];

// 由持仓原始数据 + 行情价格构建带盈亏、权重的持仓行。
export function buildHoldings(items, prices) {
  const rows = items.map(p => {
    const quote = prices[p.code];
    const price = typeof quote === "number" ? quote : quote?.price;
    const shares = Number(p.shares || 0);
    const cost = Number(p.cost || 0);
    const hasCost = cost > 0;
    const hasPrice = Number(price || 0) > 0;
    const costValue = hasCost ? shares * cost : 0;
    const marketValue = shares * Number(price || (hasCost ? cost : 0));
    const pnl = hasCost && hasPrice ? marketValue - costValue : null;
    const pnlPct = pnl == null || !costValue ? null : (pnl / costValue) * 100;
    return { ...p, shares, cost, hasCost, hasPrice, market: quote?.market || p.market, price, quoteSource: quote?.sourceLabel, costValue, marketValue, pnl, pnlPct };
  });
  const totalMarket = rows.reduce((sum, row) => sum + row.marketValue, 0);
  return rows.map(row => ({ ...row, weight: totalMarket ? (row.marketValue / totalMarket) * 100 : 0 }));
}

export function sortHoldings(rows, sort) {
  if (!sort || sort.key === "default") return rows;
  const direction = sort.direction === "asc" ? 1 : -1;
  return [...rows].sort((a, b) => {
    const av = sortableHoldingValue(a, sort.key);
    const bv = sortableHoldingValue(b, sort.key);
    if (av == null && bv == null) return 0;
    if (av == null) return 1;
    if (bv == null) return -1;
    return (av - bv) * direction;
  });
}

function sortableHoldingValue(row, key) {
  if (key === "marketValue") return Number(row.marketValue || 0);
  if (key === "pnlPct") return row.pnlPct == null ? null : Number(row.pnlPct);
  return null;
}

export function getOverview(holdings, stockRows, etfs) {
  const cost = holdings.reduce((sum, row) => sum + row.costValue, 0);
  const marketValue = holdings.reduce((sum, row) => sum + row.marketValue, 0);
  const costedMarketValue = holdings.filter(row => row.hasCost).reduce((sum, row) => sum + row.marketValue, 0);
  const pnl = costedMarketValue - cost;
  const analyzingCount = [...holdings, ...stockRows].filter(row => ["analyzing", "failed"].includes(row.analysisStatus)).length;
  const highRiskCount = [...holdings, ...stockRows].filter(row => riskLevel(row.risk) === "high").length;
  return {
    marketValue,
    pnl,
    pnlPct: cost ? (pnl / cost) * 100 : 0,
    analyzingCount,
    highRiskCount,
    positionCount: holdings.length,
    stockCount: stockRows.length,
    etfCount: etfs.length
  };
}

// PLACEHOLDER_ANALYSIS
export function buildPortfolioAnalysis(holdings) {
  const totalMarket = holdings.reduce((sum, row) => sum + Number(row.marketValue || 0), 0);
  const totalCost = holdings.reduce((sum, row) => sum + Number(row.costValue || 0), 0);
  const costedMarketValue = holdings.filter(row => row.hasCost).reduce((sum, row) => sum + Number(row.marketValue || 0), 0);
  const pnl = costedMarketValue - totalCost;
  const pnlPct = totalCost ? (pnl / totalCost) * 100 : 0;
  const count = holdings.length;
  const sortedByValue = [...holdings].sort((a, b) => Number(b.marketValue || 0) - Number(a.marketValue || 0));
  const largestHolding = sortedByValue[0] || null;
  const maxWeight = largestHolding?.weight || 0;
  const top5Weight = totalMarket ? sortedByValue.slice(0, 5).reduce((sum, row) => sum + Number(row.marketValue || 0), 0) / totalMarket * 100 : 0;
  const priceCoverage = count ? holdings.filter(row => row.hasPrice).length / count * 100 : 0;
  const costCoverage = count ? holdings.filter(row => row.hasCost).length / count * 100 : 0;
  const highRiskRows = holdings.filter(row => riskLevel(row.risk) === "high");
  const highRiskValue = highRiskRows.reduce((sum, row) => sum + Number(row.marketValue || 0), 0);
  const highRiskWeight = totalMarket ? highRiskValue / totalMarket * 100 : 0;
  const marketRows = groupHoldingRows(holdings, classifyMarketBucket, totalMarket);
  const assetRows = groupHoldingRows(holdings, classifyAssetBucket, totalMarket);
  const riskRows = groupHoldingRows(holdings, riskBucket, totalMarket);
  const themeRows = buildThemeRows(holdings, totalMarket);
  const themeKnownValue = themeRows.filter(row => row.label !== "其他/待穿透").reduce((sum, row) => sum + row.value, 0);
  const themeCoverage = totalMarket ? themeKnownValue / totalMarket * 100 : 0;
  const health = buildPortfolioHealth({ maxWeight, top5Weight, highRiskWeight, priceCoverage, costCoverage, themeCoverage });

  return {
    count,
    totalMarket,
    totalCost,
    pnl,
    pnlPct,
    largestHolding,
    maxWeight,
    top5Weight,
    topMarketWeight: marketRows[0]?.weight || 0,
    highRiskCount: highRiskRows.length,
    highRiskWeight,
    priceCoverage,
    costCoverage,
    themeCoverage,
    marketRows,
    assetRows,
    riskRows,
    themeRows,
    topAssetWeight: assetRows[0]?.weight || 0,
    pnlRows: buildAttributionRows(holdings),
    ...health
  };
}

function groupHoldingRows(holdings, getLabel, totalMarket) {
  const map = new Map();
  holdings.forEach(row => {
    const label = getLabel(row);
    const current = map.get(label) || { label, value: 0, count: 0 };
    current.value += Number(row.marketValue || 0);
    current.count += 1;
    map.set(label, current);
  });
  return Array.from(map.values())
    .map(row => ({ ...row, weight: totalMarket ? row.value / totalMarket * 100 : 0 }))
    .sort((a, b) => b.value - a.value);
}

function classifyMarketBucket(row) {
  const text = `${row.name || ""} ${row.code || ""} ${row.market || ""}`.toLowerCase();
  if (/债|货币|现金|增利|短债|纯债/.test(text)) return "固收/现金";
  if (/港股|恒生|香港|h股/.test(text)) return "港股";
  if (/美股|纳斯达克|标普|sp500|s&p|qdii|全球|海外|美元/.test(text)) return "美股/海外";
  if (/科创|半导体|芯片|集成电路|创业/.test(text)) return "A股科创成长";
  if (/中证|沪深|上证|深证|创业板|a股|etf|基金/.test(text)) return "A股宽基/基金";
  return row.market || "其他";
}

function classifyAssetBucket(row) {
  const text = `${row.name || ""} ${row.code || ""} ${row.market || ""}`.toLowerCase();
  if (/债|货币|现金|增利|短债|纯债/.test(text)) return "固收基金";
  if (/qdii|全球|海外|纳斯达克|标普/.test(text)) return "QDII / 海外基金";
  if (/etf|联接|指数|中证|沪深|上证|深证|创业板|科创/.test(text)) return "指数 / ETF";
  if (/基金|混合|股票型/.test(text)) return "主动基金";
  if (/a股|港股|美股|股票/.test(text)) return "股票";
  return row.market || "其他";
}

function riskBucket(row) {
  const level = riskLevel(row.risk);
  if (level === "high") return "高风险";
  if (level === "medium") return "中风险";
  return "低风险";
}

function buildAttributionRows(holdings) {
  return holdings
    .filter(row => row.pnl != null)
    .sort((a, b) => Math.abs(Number(b.pnl || 0)) - Math.abs(Number(a.pnl || 0)))
    .slice(0, 6)
    .map(row => ({
      key: row.id || row.code,
      label: row.name,
      value: Number(row.pnl || 0),
      detail: formatSignedPct(row.pnlPct),
      tone: Number(row.pnl || 0) >= 0 ? "up" : "down"
    }));
}
// PLACEHOLDER_THEME
function buildThemeRows(holdings, totalMarket) {
  const map = new Map();
  holdings.forEach(row => {
    const themes = inferThemes(row);
    const totalWeight = themes.reduce((sum, item) => sum + item.weight, 0) || 1;
    themes.forEach(theme => {
      const value = Number(row.marketValue || 0) * theme.weight / totalWeight;
      const current = map.get(theme.label) || { label: theme.label, value: 0, contributors: [] };
      current.value += value;
      current.contributors.push({ name: row.name, value });
      map.set(theme.label, current);
    });
  });

  return Array.from(map.values())
    .map(row => ({
      ...row,
      contributors: row.contributors.sort((a, b) => b.value - a.value),
      weight: totalMarket ? row.value / totalMarket * 100 : 0
    }))
    .sort((a, b) => b.value - a.value)
    .slice(0, 10);
}

function inferThemes(row) {
  const text = `${row.name || ""} ${row.code || ""} ${row.market || ""}`.toLowerCase();
  const themes = [];
  const add = (label, weight) => themes.push({ label, weight });

  if (/光模块|光通信|cpo|光器件/.test(text)) add("光模块/CPO", 1);
  if (/dram|hbm|内存/.test(text)) add("DRAM/HBM", 1);
  if (/nand|存储|闪存/.test(text)) add("NAND/存储", 1);
  if (/半导体|芯片|集成电路/.test(text)) {
    add("芯片/半导体", 0.58);
    add("AI 算力/科技", 0.18);
    add("DRAM/HBM", 0.08);
    add("NAND/存储", 0.06);
    add("科创成长", 0.1);
  }
  if (/科创|创业/.test(text)) {
    add("科创成长", 0.46);
    add("芯片/半导体", 0.24);
    add("AI 算力/科技", 0.18);
    add("高端制造", 0.12);
  }
  if (/纳斯达克|nasdaq|全球科技|科技先锋/.test(text)) {
    add("美股科技", 0.48);
    add("AI 算力/科技", 0.27);
    add("芯片/半导体", 0.14);
    add("海外资产", 0.11);
  }
  if (/标普|sp500|s&p/.test(text)) {
    add("美股宽基", 0.7);
    add("海外资产", 0.2);
    add("AI 算力/科技", 0.1);
  }
  if (/中证500|500指数/.test(text)) add("A股中盘宽基", 1);
  if (/上证50|沪深300|中证1000|全a|a500|深证100/.test(text)) add("A股宽基", 1);
  if (/债|货币|现金|增利|短债|纯债/.test(text)) add("固收/现金", 1);
  if (/医药|医疗|创新药/.test(text)) add("医药医疗", 1);
  if (/消费|白酒|食品/.test(text)) add("消费", 1);

  return themes.length ? themes : [{ label: "其他/待穿透", weight: 1 }];
}

function buildPortfolioHealth({ maxWeight, top5Weight, highRiskWeight, priceCoverage, costCoverage, themeCoverage }) {
  let score = 100;
  if (maxWeight > 35) score -= Math.min(22, (maxWeight - 35) * 0.8);
  if (top5Weight > 80) score -= Math.min(18, (top5Weight - 80) * 0.8);
  if (highRiskWeight > 35) score -= Math.min(20, (highRiskWeight - 35) * 0.7);
  if (priceCoverage < 90) score -= Math.min(18, (90 - priceCoverage) * 0.5);
  if (costCoverage < 90) score -= Math.min(14, (90 - costCoverage) * 0.35);
  if (themeCoverage < 55) score -= Math.min(12, (55 - themeCoverage) * 0.25);
  const healthScore = Math.max(0, Math.round(score));
  const healthTone = healthScore >= 80 ? "good" : healthScore >= 60 ? "warn" : "bad";
  const healthLabel = healthScore >= 80 ? "结构稳健" : healthScore >= 60 ? "需要复核" : "风险偏高";
  const healthAlerts = [];

  if (maxWeight > 35) healthAlerts.push({ text: `最大单仓 ${formatPercent(maxWeight)}，集中度偏高`, tone: "warn" });
  if (top5Weight > 80) healthAlerts.push({ text: `前五持仓 ${formatPercent(top5Weight)}，组合分散度不足`, tone: "warn" });
  if (highRiskWeight > 35) healthAlerts.push({ text: `高风险仓位 ${formatPercent(highRiskWeight)}，需复核止损线`, tone: "bad" });
  if (priceCoverage < 90) healthAlerts.push({ text: `行情覆盖 ${formatPercent(priceCoverage)}，部分市值待更新`, tone: "warn" });
  if (costCoverage < 90) healthAlerts.push({ text: `成本覆盖 ${formatPercent(costCoverage)}，收益归因不完整`, tone: "warn" });
  if (themeCoverage < 55) healthAlerts.push({ text: `主题识别 ${formatPercent(themeCoverage)}，底仓穿透待增强`, tone: "muted" });
  if (!healthAlerts.length) healthAlerts.push({ text: "仓位、风险和数据覆盖暂无明显异常", tone: "good" });

  return {
    healthScore,
    healthTone,
    healthLabel,
    healthAlerts,
    healthFactors: [
      { label: "最大单仓", value: formatPercent(maxWeight), percent: Math.min(100, maxWeight) },
      { label: "前五集中", value: formatPercent(top5Weight), percent: Math.min(100, top5Weight) },
      { label: "高风险仓位", value: formatPercent(highRiskWeight), percent: Math.min(100, highRiskWeight) },
      { label: "数据覆盖", value: formatPercent(Math.min(priceCoverage, costCoverage)), percent: Math.min(priceCoverage, costCoverage) },
    ]
  };
}
// PLACEHOLDER_RISK
export function riskLevel(text = "") {
  if (!text) return "low";
  if (/止损|跌破|失效|威胁|高风险|替代|下调|回调/.test(text)) return "high";
  if (/波动|不及预期|需求|政策|竞争|估值/.test(text)) return "medium";
  return "low";
}

export function actionLabel(text = "", status = "") {
  if (status === "analyzing") return "分析中";
  if (status === "failed") return "待重试";
  if (/止损/.test(text)) return "止损";
  if (/止盈/.test(text)) return "止盈";
  if (/减仓/.test(text)) return "减仓";
  if (/加仓/.test(text)) return "加仓";
  if (/持有/.test(text)) return "持有";
  if (/观察|关注/.test(text)) return "观察";
  return text ? "待复核" : "待分析";
}

export function statusText(status) {
  if (status === "done") return "已分析";
  if (status === "analyzing") return "分析中";
  if (status === "failed") return "失败";
  return "待分析";
}

File: backend/app/services/portfolio_analysis.py
"""组合分析纯计算（移植 src/lib/portfolio-analysis.js，方案 §11.4 补齐）。

buildHoldings：持仓 + 行情 → 带盈亏/权重的持仓行。
buildAnalysis：市场/资产/风险分布 + 收益归因 + 健康度 + 底层主题穿透。
全程纯函数、无 I/O，便于单测；口径与旧 JS 版严格对齐。
"""

from __future__ import annotations

import re
from typing import Any

CHART_COLORS = ["#2563eb", "#0f766e", "#dc2626", "#7c3aed", "#d97706", "#0891b2", "#64748b", "#be185d"]


def risk_level(text: str | None = "") -> str:
    t = (text or "").strip()
    if not t:
        return "low"
    if re.search(r"低风险|风险可控|相对稳健|暂无明显风险", t):
        return "low"
    if re.search(r"高风险|极高风险|重大风险|退市|爆仓|暴雷|资不抵债|流动性枯竭", t):
        return "high"
    if re.search(r"止损|跌破|失效|威胁|替代|下调|回调|波动|不及预期|需求|政策|竞争|估值", t):
        return "medium"
    return "low"


def build_holdings(rows: list[dict[str, Any]], quotes: dict[str, dict[str, Any]]) -> list[dict[str, Any]]:
    """持仓行 + 行情 → 带 marketValue/pnl/pnlPct/weight 的行（移植 buildHoldings）。"""
    out: list[dict[str, Any]] = []
    for p in rows:
        quote = quotes.get(p["code"])
        price = float(quote["price"]) if quote and quote.get("price") else None
        shares = float(p.get("shares") or 0)
        cost = float(p.get("cost") or 0)
        has_cost = cost > 0
        has_price = bool(price and price > 0)
        cost_value = shares * cost if has_cost else 0.0
        market_value = shares * (price if price else (cost if has_cost else 0.0))
        pnl = (market_value - cost_value) if (has_cost and has_price) else None
        pnl_pct = (pnl / cost_value * 100) if (pnl is not None and cost_value) else None
        out.append(
            {
                **p,
                "shares": shares,
                "cost": cost,
                "hasCost": has_cost,
                "hasPrice": has_price,
                "market": (quote.get("market") if quote else None) or p.get("market") or "",
                "price": price,
                "changePct": quote.get("changePct") if quote else None,
                "quoteSource": quote.get("sourceLabel") if quote else None,
                "costValue": cost_value,
                "marketValue": market_value,
                "pnl": pnl,
                "pnlPct": pnl_pct,
            }
        )
    total = sum(r["marketValue"] for r in out)
    for r in out:
        r["weight"] = (r["marketValue"] / total * 100) if total else 0.0
    return out


def build_overview(holdings: list[dict[str, Any]]) -> dict[str, Any]:
    """概览：总市值/浮动盈亏/AI 待处理/高风险（移植 getOverview，持仓维度）。"""
    cost = sum(r["costValue"] for r in holdings)
    market_value = sum(r["marketValue"] for r in holdings)
    costed_mv = sum(r["marketValue"] for r in holdings if r["hasCost"])
    pnl = costed_mv - cost
    analyzing = sum(1 for r in holdings if r.get("analysisStatus") in ("analyzing", "failed"))
    high_risk = sum(1 for r in holdings if risk_level(r.get("risk")) == "high")
    return {
        "marketValue": market_value,
        "pnl": pnl,
        "pnlPct": (pnl / cost * 100) if cost else 0.0,
        "analyzingCount": analyzing,
        "highRiskCount": high_risk,
        "positionCount": len(holdings),
    }


def _classify_market(row: dict[str, Any]) -> str:
    text = f"{row.get('name', '')} {row.get('code', '')} {row.get('market', '')}".lower()
    if re.search(r"债|货币|现金|增利|短债|纯债", text):
        return "固收/现金"
    if re.search(r"港股|恒生|香港|h股", text):
        return "港股"
    if re.search(r"美股|纳斯达克|标普|sp500|s&p|qdii|全球|海外|美元", text):
        return "美股/海外"
    if re.search(r"科创|半导体|芯片|集成电路|创业", text):
        return "A股科创成长"
    if re.search(r"中证|沪深|上证|深证|创业板|a股|etf|基金", text):
        return "A股宽基/基金"
    return row.get("market") or "其他"


def _classify_asset(row: dict[str, Any]) -> str:
    text = f"{row.get('name', '')} {row.get('code', '')} {row.get('market', '')}".lower()
    if re.search(r"债|货币|现金|增利|短债|纯债", text):
        return "固收基金"
    if re.search(r"qdii|全球|海外|纳斯达克|标普", text):
        return "QDII / 海外基金"
    if re.search(r"etf|联接|指数|中证|沪深|上证|深证|创业板|科创", text):
        return "指数 / ETF"
    if re.search(r"基金|混合|股票型", text):
        return "主动基金"
    if re.search(r"a股|港股|美股|股票", text):
        return "股票"
    return row.get("market") or "其他"


def _risk_bucket(row: dict[str, Any]) -> str:
    return {"high": "高风险", "medium": "中风险", "low": "低风险"}[risk_level(row.get("risk"))]


def _group_rows(holdings: list[dict[str, Any]], get_label: Any, total: float) -> list[dict[str, Any]]:
    agg: dict[str, dict[str, Any]] = {}
    for row in holdings:
        label = get_label(row)
        cur = agg.setdefault(label, {"label": label, "value": 0.0, "count": 0})
        cur["value"] += row["marketValue"]
        cur["count"] += 1
    rows = [{**r, "weight": (r["value"] / total * 100) if total else 0.0} for r in agg.values()]
    return sorted(rows, key=lambda r: r["value"], reverse=True)


# ---- 底层主题穿透（规则估算，移植 inferThemes/buildThemeRows）----


def _infer_themes(row: dict[str, Any]) -> list[tuple[str, float]]:
    text = f"{row.get('name', '')} {row.get('code', '')} {row.get('market', '')}".lower()
    themes: list[tuple[str, float]] = []
    if re.search(r"光模块|光通信|cpo|光器件", text):
        themes.append(("光模块/CPO", 1.0))
    if re.search(r"dram|hbm|内存", text):
        themes.append(("DRAM/HBM", 1.0))
    if re.search(r"nand|存储|闪存", text):
        themes.append(("NAND/存储", 1.0))
    if re.search(r"半导体|芯片|集成电路", text):
        themes += [
            ("芯片/半导体", 0.58),
            ("AI 算力/科技", 0.18),
            ("DRAM/HBM", 0.08),
            ("NAND/存储", 0.06),
            ("科创成长", 0.1),
        ]
    if re.search(r"科创|创业", text):
        themes += [("科创成长", 0.46), ("芯片/半导体", 0.24), ("AI 算力/科技", 0.18), ("高端制造", 0.12)]
    if re.search(r"纳斯达克|nasdaq|全球科技|科技先锋", text):
        themes += [("美股科技", 0.48), ("AI 算力/科技", 0.27), ("芯片/半导体", 0.14), ("海外资产", 0.11)]
    if re.search(r"标普|sp500|s&p", text):
        themes += [("美股宽基", 0.7), ("海外资产", 0.2), ("AI 算力/科技", 0.1)]
    if re.search(r"中证500|500指数", text):
        themes.append(("A股中盘宽基", 1.0))
    if re.search(r"上证50|沪深300|中证1000|全a|a500|深证100", text):
        themes.append(("A股宽基", 1.0))
    if re.search(r"债|货币|现金|增利|短债|纯债", text):
        themes.append(("固收/现金", 1.0))
    if re.search(r"医药|医疗|创新药", text):
        themes.append(("医药医疗", 1.0))
    if re.search(r"消费|白酒|食品", text):
        themes.append(("消费", 1.0))
    return themes or [("其他/待穿透", 1.0)]


def _build_theme_rows(holdings: list[dict[str, Any]], total: float) -> list[dict[str, Any]]:
    agg: dict[str, dict[str, Any]] = {}
    for row in holdings:
        themes = _infer_themes(row)
        weight_sum = sum(w for _, w in themes) or 1.0
        for label, w in themes:
            value = row["marketValue"] * w / weight_sum
            cur = agg.setdefault(label, {"label": label, "value": 0.0, "contributors": []})
            cur["value"] += value
            cur["contributors"].append({"name": row.get("name"), "value": value})
    rows = []
    for r in agg.values():
        r["contributors"] = sorted(r["contributors"], key=lambda c: c["value"], reverse=True)
        r["weight"] = (r["value"] / total * 100) if total else 0.0
        rows.append(r)
    return sorted(rows, key=lambda r: r["value"], reverse=True)[:10]


def _build_attribution(holdings: list[dict[str, Any]]) -> list[dict[str, Any]]:
    rows = [r for r in holdings if r["pnl"] is not None]
    rows.sort(key=lambda r: abs(r["pnl"]), reverse=True)
    return [
        {"label": r.get("name"), "value": r["pnl"], "detailPct": r["pnlPct"], "tone": "up" if r["pnl"] >= 0 else "down"}
        for r in rows[:6]
    ]


def _build_health(
    max_weight: float, top5: float, high_risk_w: float, price_cov: float, cost_cov: float, theme_cov: float
) -> dict[str, Any]:
    score = 100.0
    if max_weight > 35:
        score -= min(22, (max_weight - 35) * 0.8)
    if top5 > 80:
        score -= min(18, (top5 - 80) * 0.8)
    if high_risk_w > 35:
        score -= min(20, (high_risk_w - 35) * 0.7)
    if price_cov < 90:
        score -= min(18, (90 - price_cov) * 0.5)
    if cost_cov < 90:
        score -= min(14, (90 - cost_cov) * 0.35)
    if theme_cov < 55:
        score -= min(12, (55 - theme_cov) * 0.25)
    health = max(0, round(score))
    tone = "good" if health >= 80 else "warn" if health >= 60 else "bad"
    label = "结构稳健" if health >= 80 else "需要复核" if health >= 60 else "风险偏高"
    alerts: list[dict[str, str]] = []
    if max_weight > 35:
        alerts.append({"text": f"最大单仓 {max_weight:.1f}%，集中度偏高", "tone": "warn"})
    if top5 > 80:
        alerts.append({"text": f"前五持仓 {top5:.1f}%，组合分散度不足", "tone": "warn"})
    if high_risk_w > 35:
        alerts.append({"text": f"高风险仓位 {high_risk_w:.1f}%，需复核止损线", "tone": "bad"})
    if price_cov < 90:
        alerts.append({"text": f"行情覆盖 {price_cov:.1f}%，部分市值待更新", "tone": "warn"})
    if cost_cov < 90:
        alerts.append({"text": f"成本覆盖 {cost_cov:.1f}%，收益归因不完整", "tone": "warn"})
    if theme_cov < 55:
        alerts.append({"text": f"主题识别 {theme_cov:.1f}%，底仓穿透待增强", "tone": "muted"})
    if not alerts:
        alerts.append({"text": "仓位、风险和数据覆盖暂无明显异常", "tone": "good"})
    return {
        "healthScore": health,
        "healthTone": tone,
        "healthLabel": label,
        "healthAlerts": alerts,
        "healthFactors": [
            {"label": "最大单仓", "value": max_weight, "percent": min(100, max_weight)},
            {"label": "前五集中", "value": top5, "percent": min(100, top5)},
            {"label": "高风险仓位", "value": high_risk_w, "percent": min(100, high_risk_w)},
            {"label": "数据覆盖", "value": min(price_cov, cost_cov), "percent": min(price_cov, cost_cov)},
        ],
    }


def build_analysis(holdings: list[dict[str, Any]]) -> dict[str, Any]:
    """组合分析（移植 buildPortfolioAnalysis）：分布 + 归因 + 健康度 + 主题穿透。"""
    total = sum(r["marketValue"] for r in holdings)
    total_cost = sum(r["costValue"] for r in holdings)
    costed_mv = sum(r["marketValue"] for r in holdings if r["hasCost"])
    pnl = costed_mv - total_cost
    count = len(holdings)
    if not count:
        return {"count": 0}
    by_value = sorted(holdings, key=lambda r: r["marketValue"], reverse=True)
    largest = by_value[0]
    max_weight = largest.get("weight", 0)
    top5 = (sum(r["marketValue"] for r in by_value[:5]) / total * 100) if total else 0.0
    price_cov = sum(1 for r in holdings if r["hasPrice"]) / count * 100
    cost_cov = sum(1 for r in holdings if r["hasCost"]) / count * 100
    high_risk = [r for r in holdings if risk_level(r.get("risk")) == "high"]
    high_risk_w = (sum(r["marketValue"] for r in high_risk) / total * 100) if total else 0.0
    market_rows = _group_rows(holdings, _classify_market, total)
    asset_rows = _group_rows(holdings, _classify_asset, total)
    risk_rows = _group_rows(holdings, _risk_bucket, total)
    theme_rows = _build_theme_rows(holdings, total)
    theme_known = sum(r["value"] for r in theme_rows if r["label"] != "其他/待穿透")
    theme_cov = (theme_known / total * 100) if total else 0.0
    health = _build_health(max_weight, top5, high_risk_w, price_cov, cost_cov, theme_cov)
    return {
        "count": count,
        "totalMarket": total,
        "totalCost": total_cost,
        "pnl": pnl,
        "pnlPct": (pnl / total_cost * 100) if total_cost else 0.0,
        "largestHolding": {"name": largest.get("name"), "weight": max_weight},
        "maxWeight": max_weight,
        "top5Weight": top5,
        "topMarketWeight": market_rows[0]["weight"] if market_rows else 0.0,
        "topAssetWeight": asset_rows[0]["weight"] if asset_rows else 0.0,
        "highRiskCount": len(high_risk),
        "highRiskWeight": high_risk_w,
        "priceCoverage": price_cov,
        "costCoverage": cost_cov,
        "themeCoverage": theme_cov,
        "marketRows": market_rows,
        "assetRows": asset_rows,
        "riskRows": risk_rows,
        "themeRows": theme_rows,
        "pnlRows": _build_attribution(holdings),
        **health,
    }
